Refactor greedy pauli simp

yao-cqc opened the pull request:
# Description

Supports mid-circuit measurements, resets, conditionals and classical gates.

A modified version of the existing Pauli graph is needed for this feature.

# Related issues

Please mention any github issues addressed by this PR.

# Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

sjdilkes reviewed (changes requested):
Code looks good, just a question on functionality!

yao-cqc commented:
Several changes since the last review:

Conditionals are now added as conditional _blocks_ of PauliRotations, and any two blocks can be merged if they are adjacent and have the same conditions. Each block will be synthesised recursively using GreedyPauliSimp when it's ready to be added to the output circuit.

Randomised tie breaking when selecting TQEs

`max_tqe_candidates` controls the maximum number of TQE gates to be evaluated at each step. If more TQEs than this number then a set of size `max_tqe_candidates` will be randomly sampled.

`max_lookahead` controls the number of nodes to be costed when evaluating a TQE.

If `allow_zzphase` is true, then ZZPhase gates and TQE gates will all be considered when selecting an action. A ZZPhase will reduce the tqe cost of one 2-q rotation node by 1 and has no affect on other nodes.

yao-cqc commented on tket/include/tket/Transformations/GreedyPauliOptimisation.hpp:
```diff
@@ -56,148 +88,467 @@ enum class LocalCliffordType {
+/**
+ * @brief Black box node for classical Ops
+ */
+class ClassicalNode : public PauliNode {
```
I have added a python test to check all of the above. Only Barrier is not supported.

sjdilkes reviewed (dismissed):
Works for me, nice!
